src/coreComponents/meshUtilities/CPMesh/CPMeshComm.cpp

```diff
@@ -0,0 +1,142 @@
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file CPMeshComm.cpp
+ */
+
+#include "CPMeshComm.hpp"
+
+#include "mpiCommunications/MpiWrapper.hpp"
+
+namespace geosx
+{
+
+namespace CPMesh
+{
+
+CPMeshComm::CPMeshComm( string const & name )
+  :
+  m_mainRank( 0 ),
+  m_myRank( -1 ),
+  m_size( 0 ),
+  m_meshName( name )
+{}
+
+void CPMeshComm::setupMPIPartition( CPMeshDimensions & meshDims )
+{
+  m_myRank = MpiWrapper::commRank( MPI_COMM_GEOSX );
+  m_size = MpiWrapper::commSize( MPI_COMM_GEOSX );
+
+  if( m_size == 1 ) // serial run
+  {
+    meshDims.definePartitionBoundaries( 0, 0, meshDims.nX()-1, meshDims.nY()-1 );
+    meshDims.definePartitionOverlaps( 0, 0, 0, 0 );
+  }
+  else // parallel run
+  {
+    broadcastDomainDimensions( meshDims );
+    scatterPartitionBoundaries( meshDims );
+  }
+}
+
+void CPMeshComm::broadcastDomainDimensions( CPMeshDimensions & meshDims ) const
+{
+  localIndex nCellsInEachDirection[3];
+
+  // the main rank has read the top of the file and knows the domain dimensions
+  if( m_myRank == m_mainRank )
+  {
+    nCellsInEachDirection[0] = meshDims.nX();
+    nCellsInEachDirection[1] = meshDims.nY();
+    nCellsInEachDirection[2] = meshDims.nZ();
+  }
+  localIndex const sendRecvSize = 3;
+  MpiWrapper::bcast( nCellsInEachDirection, sendRecvSize, m_mainRank, MPI_COMM_GEOSX );
+  MpiWrapper::barrier();
+
+  // the other ranks have received the data and can set these values in CPMeshData
+  if( m_myRank != m_mainRank )
+  {
+    meshDims.defineDomainDimensions( nCellsInEachDirection[0],
+                                     nCellsInEachDirection[1],
+                                     nCellsInEachDirection[2] );
+  }
+}
+
+void CPMeshComm::scatterPartitionBoundaries( CPMeshDimensions & meshDims )
+{
+  localIndex const sizePerRank = 4;
+  array1d< localIndex > boundaries( sizePerRank*m_size );
+  array1d< localIndex > overlaps( sizePerRank*m_size );
+  array1d< localIndex > neighbors( sizePerRank*m_size );
+
+  if( m_myRank == m_mainRank )
+  {
+    // Temporary code to facilitate debugging: 1D partitioning along the x-direction.
+    // Ultimately here we will partition in both X and Y, taking into account active cells in each partition.
+    // Once I am sure that everything works on large test case, I will rewrite this entirely
+    localIndex const spacing = floor( static_cast< real64 >( meshDims.nX() )
+                                      / static_cast< real64 >( m_size ) );
+    for( localIndex iRank = 0; iRank < m_size; ++iRank )
+    {
+      boundaries( iRank*sizePerRank )   = iRank * spacing;
+      boundaries( iRank*sizePerRank+1 ) = 0;
+      boundaries( iRank*sizePerRank+2 ) = ( iRank < m_size-1 ) ? ( iRank+1 ) * spacing - 1 : meshDims.nZ()-1;
+      boundaries( iRank*sizePerRank+3 ) = meshDims.nY()-1;
```

Comment:
For this comment and most of the others in this class, I would prefer not doing anything for the moment. Currently the partitioning is Cartesian. Each "interior" rank has four neighbors, and could in fact deduce the dimensions of its domain as well as the indices of the neighbors without doing any communication (no scatter), as you remarked. 

For domains with a lot of inactive cells like Norne or SAIGUP, this is not going to work very well because many ranks will not have any active cell. For these cases, we will have to abandon the strictly Cartesian approach, and we will have to introduce a little bit more flexibility in the partitioning. At that point, the dimensions of each partition will have to be communicated and many things in this function will have to change (`sizePerRank`, the neighbor indices, etc).